### ISSUE TITLE
[RFC] Evolution of Backstage core and backwards compatibility of plugins

### ISSUE THREAD

**Issue by Rugvip** [backstage/backstage#4872]
**Status:** Open for comments

**Authors:** [Rugvip](https://github.com/Rugvip), [joaomilho](https://github.com/joaomilho)

## Need

In backstage apps it is common to end up with multiple versions of `@backstage/core`. This breaks applications because of duplicated state – for example React context instances. Right now we're combating this problem by trying to avoid duplicated packages using utilities like `backstage-cli versions:check --fix`. This solution works well as long as all requested versions of `@backstage/core` are semver compatible, but as the Backstage ecosystem grows we want to allow for more flexibility in the versioning of plugins, and not require all of them to use the same version of the core library. We also want to significantly reduce the impact of a semver breaking change to the core API, as this will allow us to evolve more quickly while still providing a stable platform.

Below is an illustration of the problem that arises from having multiple plugins that depend on different versions of `@backstage/core`. The app and the plugins each depend on their own version, with none of them being compatible with each other. This results in the core package being duplicated 3 times, one in the root `node_modules`, and the other two in package local `node_modules`. At runtime this will manifest itself as errors accessing React contexts, most commonly crashing on usages of `useApi`, as well as a significantly increased bundle size.

![image](https://user-images.githubusercontent.com/4984472/110328461-aecbdb80-801b-11eb-955f-50565f2d5528.png)

## Proposal

In order to reach the desired flexibility in plugin versioning, we plan to add two new solutions to our versioning strategy, 1) avoid the occurrence of duplication of context _providers_ while isolating the occurrence of duplications of context _consumers_, 2) allow these duplications to work.

### Isolate duplications

To decrease duplications, we need to reduce the surface area of the library that provides the contexts. We plan to do that by splitting the `@backstage/core` library in 3 parts:

- Provider: A library that contains only the context providers (createApp) - `@backstage/app-api` (?)
- Consumer: A library that contains only the context consumers (e.g. useApi) - `@backstage/plugin-api` (?)
- Components: A library that contains what's left from core, mainly components - `@backstage/components` (?)

This way, the provider library will only exist _once_, since it's used only when creating apps. This will guarantee the provider is the one provided by the version of the library that is installed as part of packages/app.

This split will also decrease app bundle sizes, since there's no duplication of the provider package.

### Allowing duplications to work

Duplications of the consumer library will still happen, since they are used by plugins and the application itself. To solve this problem we introduce a global shared state of contexts singletons. The global singletons introduce another problem however, which is that we're locked to a single interface. To solve that, we make the values that the singletons store versioned, and then let consumers ask for the version that they understand. Context providers will become a bit more complex, since they will have to provide all of its allowed versions. The upsides are that:

- Plugins don't necessarily need to be updated when Backstage is updated.
- Apps will be able to include older plugins.
- Open source plugins will also be able to provide versioned utility APIs.

We would still not solve the problem of plugins being _ahead_ of the providers (them being backstage or other plugins). We believe it is best to force upgrades in these cases. This could be solved by tooling and also by typescript (the consumer would be trying to get a version that is not part of the provided versions). In any case, this solution _could_ also be applied to that problem if needed although forwards compatibility would add even more complexity.

### Implementation

We provide implementation examples for creating packages and version boundaries across two different systems, React contexts, and Utility APIs. They are both based around a core `VersionedValue` type, which is a container for accessing values at multiple different versions.

```ts
interface VersionedValue<Versions extends { [version in number]: any }> {
  atVersion<Version extends keyof Versions>(version: Version): Versions[Version];
}
```

#### React Context

The most common and important piece of state that needs to be communicated across packages and version boundaries are React contexts. This sample implementation shows how we can start out with an initial version of a context, and then evolve the context by adding a second version while still maintaining backwards compatibility with consumers of the initial version.

<details>
<summary>
Sample implementation using versioned React Contexts
</summary>

##### Version 1

We start off with a simple `Greeting` context. Since we know this context is one that will be used across multiple packages we make it a versioned context from
the start.

```tsx
interface GreetingApi {
  greet(): string;
}

class Greeting implements GreetingApi {
  greet() {
    return 'hello';
  }
}

const GreetingContext = createContext<
  VersionedValue<{ 1: GreetingApi }> | undefined
>(undefined);

// In this example we use the window object to share the context, but this could be done in other ways
window.__greetingContext = GreetingContext;

const GreetingProvider = ({ children }) => {
  // createVersionedValueMap is a utility that implements { atVersion(version: number) } for us
  const versionedGreeting = createVersionedValueMap({ 1: new Greeting() });

  return (
    <GreetingContext value={versionedGreeting}>{children}</GreetingContext>
  );
};

// Plugin
const MyGreeting = () => {
  // In this case we'd access the context we need from the window object, since we're
  // consuming the context at render time we're guaranteed for it to have been set first.
  const GreetingContext = window.__greetingContext;

  const greetings = useContext(GreetingContext)?.atVersion(1);
  if (!greetings) {
    throw new Error('Greeting API v1 is not available');
  }

  return <h1>{greetings.greet()}</h1>;
};
```

##### Version 2

In this version we evolve the `Greeting` context to provide a second version that is able to greet by name. For old consumers we provide an adapter that implements version 1 of the context using version 2.

```tsx
interface GreetingApiV1 {
  greet(): string;
}

interface GreetingApiV2 {
  greet(name: string): string;
}

class Greeting implements GreetingApiV2 {
  greet(name: string) {
    return `Hello ${name}!`;
  }
}

class GreetingV1Adapter implements GreetingApiV1 {
  constructor(private readonly v2: GreetingApiV2) {}

  greet() {
    return this.v2.greet('user');
  }
}

const GreetingContext = createContext<
  VersionedValue<{ 1: GreetingApiV1; 2: GreetingApiV2 }> | undefined
>(undefined);

window.__greetingContext = GreetingContext;

const GreetingProvider = ({ children }) => {
  const greeting = new Greeting();
  const versionedGreeting = createReducedVersionedValueMap({
    1: new GreetingV1Adapter(greeting),
    2: greeting,
  });

  return (
    <GreetingContext value={versionedGreeting}>{children}</GreetingContext>
  );
};

// Old Plugin (v1)
const MyGreeting = () => {
  const GreetingContext = window.__greetingContext;
  const greetings = useContext(GreetingContext)?.atVersion(1);
  if (!greetings) {
    throw new Error('Greeting API v1 is not available');
  }

  return <h1>{greetings.greet()}</h1>;
};

// New plugin (v2)
const MyGreeting = () => {
  const GreetingContext = window.__greetingContext;
  const user = useUser();
  const greetings = useContext(GreetingContext)?.atVersion(2);
  if (!greetings) {
    throw new Error('Greeting API v2 is not available');
  }

  return <h1>{greetings.greet(user.name)}</h1>;
};
```

</details>

#### Utility APIs

Another system that we could apply the versioned value strategy to are the Utility APIs. In this case it is more important that we're able to move from an API that is not versioned, to one that is versioned. We also want to apply the versioning in a more transparent way, keeping the existing consumption patterns of Utility APIs mostly intact.

> Note that we do not propose to include this in the initial implementation, but rather might add it as a followup.

<details>
<summary>
Sample implementation of versioned Utility APIs
</summary>

##### Unversioned

We start off with a plain Utility API implementation and consumption pattern.

```tsx
const greetingApiRef = createApiRef<GreetingApi>({
  id: 'myPlugin.greeting',
  description: 'Greets the user',
});

/// usage
createApiFactory({
  api: greetingApiRef,
  deps: {},
  factory: () => new Greeting(),
});

// Plugin
const greetingApiRef = createApiRef<GreetingApi>({
  id: 'myPlugin.greeting',
});

function MyCompo() {
  const greeting = useApi(greetingApiRef);
  return <h1>{greeting.greet()}</h1>;
}
```

##### Version 2

As we want to introduce version 2 of the API, we also need to make it versioned. We do this by providing a version to the API ref. The version that we pass will be used as the max version when producing an implementation of the API, and the desired version when consuming it.

```tsx
const greetingApiRef = createApiRef<{1: GreetingApiV1, 2: GreetingApiV2}>({
  id: 'myPlugin.greeting',
  description: 'Greets the user',
  version: 2, // The version ends up being embedded as the second type parameter
});

// Defining a versioned API factory, needs to return a VersionedValue<> of the API type
createApiFactory({
  api: greetingApiRef,
  deps: {},
  factory: () => {
    const greeting = new Greeting();

    return createVersionedValueMap({
      1: new GreetingV1Adapter(greeting),
      2: greeting,
    });
  },
});

class VersionedApiHolder {
  isImplVersioned(ref: ApiRef<any>) {
    const implRef = this.implRefsById.get(ref.id) // Note that refs are identified by ID now
    return 'version' in implRef;
  }

  // The API implementation resolution process might look something like this
  get<T>(ref: ApiRef<T>): T {
    const isRefVersioned = 'version' in ref
    const isImplVersioned = this.isImplVersioned(ref)
    const impl = this.getImpl(ref);

    if (isRefVersioned && !isImplVersioned) {
      if (ref.version === 1) {
        return impl
      }
      throw new Error('NOPE')
    }
    if (!isRefVersioned && isImplVersioned) {
      return impl.atVersion(1)
    }
    if (isRefVersioned && isImplVersioned) {
      return impl.atVersion(ref.version)
    }
    return impl
  }
}

function useApi(ref) {
  const holder = useContext(ApiContext);
  return holder.get(ref);
}

// old greetingPlugin (v1)
const greetingApiRef: ApiRef<GreetingApi, undefined> = createApiRef<GreetingApi>({
  id: 'myPlugin.greeting',
});

function MyCompo() {
  const greeting = useApi(greetingApiRef);
  return <h1>{greeting.greet()}</h1>;
}

// new greetingPlugin (v2)
const greetingApiRef: ApiRef<..., 2> = createApiRef<{1: GreetingApiV1, 2: GreetingApiV2}>({
  id: 'myPlugin.greeting',
  version: 2,
});

// backstage installation
function MyCompo() {
  const greeting = useApi(greetingApiRef); // greeting will be inferred to GreetingApiV2
  return <h1>{greeting.greet()}</h1>;
}
```

</details>

### Keep `@backstage/core` stable

This would prohibit core and by extension the Backstage project from evolving.

### Force versions using Yarn resolutions

This approach would pin the version of `@backstage/core` to a single version for all plugins in a project. While this does ensure that there is only source of truth at runtime, it is also a very risky approach and one prone to breakage, especially over time and as the app size increases. The contract that is put in place between each plugin and the version of the core package that it depends on is broken, meaning that breakages will happen at runtime in ways that are also often hard to debug.

### Keep all plugins up to date

This would force all apps to upgrade when core is updated. This would generate exponentially more work towards our users, given our user base increases. This is not a realistic alternative given our expansion plans.

### Context package

The solution in this approach would be to separate out all context objects into a separate package, i.e. a package containing only `React.createContext` exports. While this approach would solve the duplication of context objects, it does not allow for evolution of the Utility APIs in the core package.

We also preferred using an approach to sharing the context objects that doesn't create extra packages, as it would again lead to tools like `backstage-cli versions:check --fix` being required to make the project work, rather than being just optimizations to reduce bundle size.

### Webpack externals / Webpack module federation

We wanted avoided any options that had dependencies on the build tooling, in order for the libraries and tooling to be able to evolve independently. There are also many other approaches to sharing state between packages that don't need to hook into the build system.

## Risks

1.  **Complexity**

    Providers have to expose multiple versions of the API implementations.

    These are written both by backstage core team as part of the default Utility APIs and plugin authors.

    Unit test will be necessary to cover all older versions as well.

2.  **Introducing additional libraries**

    There will be a period of updating consumers to reflect the new split. After that the split is expected to be intuitive.

**Comment by kaos** [comment]
First, great reasoning in this RFC.

Are there plans for including deprecation information on older versions, that could show up in build/runtime logs/console?

**Comment by Rugvip** [comment]
@kaos Thanks!

Yes, deprecations can be handled both from the app, in warning that there are consumers that are deprecated, as well as from the plugin API, which can deprecate its own methods in the regular way. Since the consumer requests its desired version via a function call we'll be able to show a deprecated warning that includes a stack trace that allows you to identify the outdated consumer.

When it comes to the transition from the existing core package to a new setup, they'll be compatible. We'll most likely have a forwards compatible version of core-api out for a while before we introduce any further changes. Warning about incompatible versions of core-api should be doable as well, but is something we'd likely eventually want to remove.

**Comment by Rugvip** [comment]
Closing this as the split was merged in #5825. Migration docs and recommendations will be up once we've been able to play around with the new packages in the main repo.
